[build] add support for EFR32 devices on Thread 1.3

lmnotran opened the pull request:
This adds support for building efr32 images for Thread 1.3

It also refactors the `make-firmware.bash` and `make-reference-release.bash` scripts to allow other platforms to be easily added.

lmnotran committed:
Add `ot-efr32` submodule

lmnotran committed:
Add ot-efr32 to build script

lmnotran committed:
Uncomment build steps

lmnotran committed:
Fix builds for efr32 openthread-1.1

lmnotran committed:
Remove efr32 from thread 1.1 and 1.2 builds

lmnotran committed:
Add Thread 1.3 to the 'build()' function

lmnotran committed:
Use environment variables to reduce args in function calls

lmnotran committed:
Add support for efrr32 to ./script/make-reference-release.bash

lmnotran committed:
Use wildcard to simplify .gitignore

lmnotran committed:
Switch ot-efr32 submodule url to HTTPS

lmnotran committed:
Fix error message

romacdon commented on README.md:
```diff
@@ -50,8 +50,6 @@ When `REFERENCE_RELEASE_TYPE` is `1.3`, reference release contains following com
 - Change log
 - Quick start guide
 
-Note: Currently, only nRF52840 dongles are supported for CLI/RCP firmwares.
```
Mason, Let's put this comment back and add to it.  We can say something like, `Note: Currently, only nRF52840 dongles and EFR32 4166A boards are supported for CLI/RCP firmwares.`

romacdon reviewed (commented):
Why are we updating the top level openthread submodule?

Also, the ot-efr32 submodule hash is from a couple of weeks ago.

romacdon commented on script/make-firmware.bash:
```diff
@@ -29,98 +29,312 @@
+platform=""
+build_dir=""
+
+readonly OT_PLATFORMS=(nrf52840 efr32mg1 efr32mg12 efr32mg13 efr32mg21)
```
This should just be efr32mg12 board 4166A

romacdon commented on script/make-firmware.bash:
```diff
@@ -29,98 +29,312 @@
+)
+
+readonly build_1_2_options_efr32=(
+    '-DOT_CSL_RECEIVER=OFF'
```
Shouldn't this be `ON`

lmnotran replied:
This actually shouldn't be here so I deleted it. 

@suveshpratapa Should I enable CSL receiver for 1.3?

suveshpratapa replied:
The OT GitHub repo has actually made a change that turns these options ON by default for stack versions > 1.2.  So I think you're good removing this option here.

romacdon commented on script/make-reference-release.bash:
```diff
@@ -29,22 +29,61 @@
 set -euxo pipefail
 
-mkdir -p build
+readonly OT_PLATFORMS=(nrf52840 efr32mg1 efr32mg12 efr32mg13 efr32mg21)
```
We should only support efr32mg12 board 4166

romacdon commented on script/make-firmware.bash:
```diff
@@ -29,98 +29,312 @@
-  local BINARY_PATH=${3:-"./build-$2/bin/$1"}
-  arm-none-eabi-objcopy -O ihex "$BINARY_PATH" "$1"-"$2".hex
-  $NRFUTIL pkg generate --debug-mode --hw-version 52 --sd-req 0 --application "$1"-"$2".hex --key-file /tmp/private.pem "$1"-"$2".zip
+readonly build_1_3_options_nrf=(
```
Why aren't there any Thread 1.3 NRF options?

lmnotran replied:
They aren't building with any turned on so I left this options list empty

https://github.com/openthread/ot-reference-release/blob/c2566b0af0ec7853e76066bdf41bf6bc1630c1ee/script/make-firmware.bash#L123-L126

suveshpratapa replied:
I'm very curious how Nordic is producing 1.3 builds without turning on the config options we turned on in build_1_3_options_efr32

Are these enabled somewhere else by default? If so is our option list redundant?

romacdon replied:
Sent an email to Jonathan asking about this.

romacdon commented on script/make-reference-release.bash:
```diff
@@ -29,22 +29,61 @@
-echo "REFERENCE_RELEASE_TYPE=${REFERENCE_RELEASE_TYPE?}"
+main()
+{
+    if [[ $# == 0 ]]; then
```
Since this script previously did not require a platform for backward compatibility we should either default to nrf52840 or build both nrf52840 and efr32mg12 when no platform is specified.  I think I'd prefer building both.

lmnotran replied:
Good point. I changed it to build both

romacdon commented on script/make-reference-release.bash:
```diff
@@ -29,22 +29,61 @@
-cp -r doc/* "$OUTPUT_ROOT"
-cp CHANGELOG.txt "$OUTPUT_ROOT"
+        efr32*)
+            OUTPUT_ROOT="$OUTPUT_ROOT"/fw_dongle/ BOARD=${BOARD?Please specify a EFR32 Board} ./script/make-firmware.bash "${platform}"
```
Hard code board 4166

lmnotran committed:
Limit efr32 to only brd4166a. Change 'make-reference-release.bash' so it builds all supported platforms

lmnotran committed:
Remove efr32 options for thread 1.1 and 1.2

lmnotran committed:
Fix README

suveshpratapa reviewed (commented):
Looks good generally, I want to confirm that 1.3 functionality / cli is present in the builds generated here.  Also I'm curious how Nordic is during 1.3 testing without passing 1.3 config options.

romacdon commented:
> Looks good generally, I want to confirm that 1.3 functionality / cli is present in the builds generated here. Also I'm curious how Nordic is during 1.3 testing without passing 1.3 config options.

Yes, thats my question too.  It might be that they haven't fully set it up for Thread 1.3 for Nordic yet.  I will ask Jonathan about it.

lmnotran committed:
Revert openthread to previous hash. Update ot-efr32

lmnotran committed:
Fix bug where OUTPUT_ROOT is unset for builds on macOS

romacdon commented on script/make-firmware.bash:
```diff
@@ -29,98 +29,303 @@
+
+main()
+{
+    if [[ $# == 0 ]]; then
```
For the same reasons we are now allowing make-reference-release.bash to be run without args I think we should do the same thing for this script.

lmnotran replied:
I disagree. This script is mainly used as a part of `./script/make-reference-release`. Having this script take in the platform as an argument allows for easier debugging/testing for when you only want to build one platform.

@suveshpratapa thoughts?

romacdon replied:
I'm ok with allowing for a platform specification. I'm just saying we shouldn't require one.

suveshpratapa replied:
@lmnotran Are you implying this script is not called anywhere else, and in its invocation always takes an argument? If so that's okay.

However, we shouldn't break any existing functionality, and if there's a use case for it to be called without an argument (therefore building for all platforms), we should keep it in.

lmnotran replied:
@suveshpratapa 
> @lmnotran Are you implying this script is not called anywhere else, and in its invocation always takes an argument? If so that's okay.

Yup! That's exactly what I'm saying. It's only called https://github.com/SiliconLabs/ot-reference-release/pull/1/files#diff-df3647cf1ac7f622c7a4f57f9762626a7e84d775dc4e7209aa7a013e85d0746dR47-R49

@romacdon 
> I'm ok with allowing for a platform specification. I'm just saying we shouldn't require one.

I can make it so that the `make-firmware.bash` builds all firmware when no args are given and only builds a specific platform if one is given.

romacdon replied:
> > I'm ok with allowing for a platform specification. I'm just saying we shouldn't require one.
> 
> I can make it so that the `make-firmware.bash` builds all firmware when no args are given and only builds a specific platform if one is given.

Great -= Thanks

lmnotran replied:
@romacdon Actually, I don't think doing that would the cleanest solution. We would need to loop through pretty much the entire body of `make-firmware.bash: main()` for each platform.

Can we just leave it as is and ask the OpenThread community for their input? From the perspective of the main use-case, this doesn't break anything. See https://github.com/SiliconLabs/ot-reference-release/blob/main/README.md

romacdon replied:
> @romacdon Actually, I don't think doing that would the cleanest solution. We would need to loop through pretty much the entire body of `make-firmware.bash: main()` for each platform.
> 
> Can we just leave it as is and ask the OpenThread community for their input? From the perspective of the main use-case, this doesn't break anything. See https://github.com/SiliconLabs/ot-reference-release/blob/main/README.md

My concern is that you are changing the interface.  Yes, make-reference-release.bash is the only script in the repo that is making use of it but you really don't know if users of the repo have external scripts invoking it.  Also, the concern about looping through the entire body of `make-firmware.bash: main()` for each platform I understand but you can easily solve that by moving the content of main to a function and looping through each platform in main invoking the new function.

lmnotran committed:
Add support for downloading nrfutil on MacOS

lmnotran commented:
A separate question that is probably some thing we'll need to ask Nordic or Google folks is:  how should `OUTPUT_ROOT` be structured now that multiple platforms are being built? 
The way things are in my branch, the platform name is appended to `fw_dongle`.
```shell
$ tree build           
build
└── ot-1.3-20210806-de53d721d
    ├── fw_dongle_efr32mg12
    │   ├── ot-cli-ftd-1.3-20210806-de53d721d.zip
    │   └── ot-rcp-1.3-20210806-de53d721d.zip
    └── fw_dongle_nrf52840
        ├── ot-cli-ftd-1.2-20210806-de53d721d.zip
        └── ot-rcp-1.2-20210806-de53d721d.zip
```
Should we be doing platform folders within `fw_dongle` instead?
```shell
build
└── ot-1.3-20210806-de53d721d
    └── fw_dongle
        ├── efr32mg12
        │   ├── ot-cli-ftd-1.3-20210806-de53d721d.zip
        │   └── ot-rcp-1.3-20210806-de53d721d.zip
        └── nrf52840
            ├── ot-cli-ftd-1.3-20210806-de53d721d.zip
            └── ot-rcp-1.3-20210806-de53d721d.zip
```

romacdon commented:
> A separate question that is probably some thing we'll need to ask Nordic or Google folks is: how should `OUTPUT_ROOT` be structured now that multiple platforms are being built?
> The way things are in my branch, the platform name is appended to `fw_dongle`.
> 
> ```shell
> $ tree build           
> build
> └── ot-1.3-20210806-de53d721d
>     ├── fw_dongle_efr32mg12
>     │   ├── ot-cli-ftd-1.3-20210806-de53d721d.zip
>     │   └── ot-rcp-1.3-20210806-de53d721d.zip
>     └── fw_dongle_nrf52840
>         ├── ot-cli-ftd-1.2-20210806-de53d721d.zip
>         └── ot-rcp-1.2-20210806-de53d721d.zip
> ```
> 
> Should we be doing platform folders within `fw_dongle` instead?
> 
> ```shell
> build
> └── ot-1.3-20210806-de53d721d
>     └── fw_dongle
>         ├── efr32mg12
>         │   ├── ot-cli-ftd-1.3-20210806-de53d721d.zip
>         │   └── ot-rcp-1.3-20210806-de53d721d.zip
>         └── nrf52840
>             ├── ot-cli-ftd-1.3-20210806-de53d721d.zip
>             └── ot-rcp-1.3-20210806-de53d721d.zip
> ```

Yeah, I don't know what's best here because there's no easy way to maintain backward compatibility with the previous output.  When we open the real PR let's tag Jonathan for thoughts on this question.  When I first reviewed the changes I was also a little concerned about using `fw_dongle` for efr32 builds because the board we are building for is really not a dongle.

suveshpratapa commented:
@lmnotran @romacdon 

I tried building from this repo and I was running into issues building our sleepy-demo apps in ot-efr32 because they are incompatible with the version of openthread packaged in this repo (they use the version of openthread in ot-efr32)

I talked to Mason and we decided that we shouldn't be building sleepy-demo apps in this repo.

romacdon commented:
> @lmnotran @romacdon
> 
> I tried building from this repo and I was running into issues building our sleepy-demo apps in ot-efr32 because they are incompatible with the version of openthread packaged in this repo (they use the version of openthread in ot-efr32)
> 
> I talked to Mason and we decided that we shouldn't be building sleepy-demo apps in this repo.

Correct.  It should only be building those things that are used in the TH.  I believe that to be the ot-cli-ftd, ot-rcp, and otbr

lmnotran commented:
openthread/ot-efr32#117 created to allow skipping of sleepy-demo apps

lmnotran committed:
Allow relative paths for OUTPUT_ROOT when running make-firmware.bash directly

lmnotran committed:
Skip Silabs apps in ot-efr32

suveshpratapa reviewed (approved):
@lmnotran @romacdon 

Verified the changes by building according to Mason's instructions and checked that 1.3 features are present.

romacdon commented:
@lmnotran I see you requested my review but I don't see any changes and one of my review comments is still open.

lmnotran committed:
Allow `make-firmware.bash` to be run without specifying a platform

If no platform is specified, the script will build all platforms. If a
platform is specified as an argument, the script will build only that
platform

Other changes:
- Remove "Thread version 1.3" support from build()

romacdon reviewed (approved):
LGTM

suveshpratapa reviewed (approved):
LGTM
